lib/segment/src/index/hnsw_index/tests/test_compact_graph_layer.rs

```diff
@@ -0,0 +1,84 @@
+use std::cmp::max;
+
+use itertools::Itertools;
+use rand::prelude::StdRng;
+use rand::SeedableRng;
+
+use crate::fixtures::index_fixtures::random_vector;
+use crate::index::hnsw_index::graph_layers::GraphLayersBase;
+use crate::index::hnsw_index::graph_layers_builder::GraphLayersBuilder;
+use crate::index::hnsw_index::graph_links::GraphLinksRam;
+use crate::index::hnsw_index::point_scorer::FilteredScorer;
+use crate::index::hnsw_index::tests::create_graph_layer_builder_fixture;
+use crate::spaces::simple::CosineMetric;
+use crate::vector_storage::ScoredPointOffset;
+
+fn search_in_builder(
+    builder: &GraphLayersBuilder,
+    top: usize,
+    ef: usize,
+    mut points_scorer: FilteredScorer,
+) -> Vec<ScoredPointOffset> {
+    let entry_point = match builder
+        .get_entry_points()
+        .get_entry_point(|point_id| points_scorer.check_point(point_id))
+    {
+        None => return vec![],
+        Some(ep) => ep,
+    };
+
+    let zero_level_entry = builder.search_entry(
+        entry_point.point_id,
+        entry_point.level,
+        0,
+        &mut points_scorer,
+    );
+
+    let nearest =
+        builder.search_on_level(zero_level_entry, 0, max(top, ef), &mut points_scorer, &[]);
+    nearest.into_iter().take(top).collect_vec()
+}
+
+#[test]
+/// Check that HNSW index with raw and compacted links gives the same results
+fn test_compact_graph_layers() {
```

Comment:
Another end-to-end test which checks, that old links structure dives exactly the same search results as the new one